docs/adr/0016-use-graphql-and-apollo-for-settings-redesign.md

```diff
@@ -0,0 +1,138 @@
+
+- Deciders: Lauren Zugai, Ben Bangert, Wil Clouser, Jody Heavener, Les Orchard, Danny Coates, Barry Chen, Vijay Budhram, Dave Justice, Alex Davis
+- Date: 2020-05-05
+
+## Context and Problem Statement
+
+The [Settings Redesign project](https://jira.mozilla.com/browse/FXA-840) will be created [as a new React application](https://github.com/mozilla/fxa/blob/master/docs/adr/0011-create-new-react-app-for-settings-redesign.md) and in turn, has opened the door to assess certain pieces of our technology stack.
+
+[GraphQL](https://graphql.org/), or GQL, is not a database query language, but is instead a query language for APIs. It _describes_ data requirements, and is a powerful alternative to REST. Some benefits can be gained by using it on top of existing REST architecture.
+
+"Apollo" in this document refers to the "Apollo client" and "Apollo server" pieces of the Apollo platform¹, which can be described as a unified data layer that enables applications to interact with data from data stores and APIs. In other words, it allows us to write and handle GraphQL on the client and server. Apollo also gives us many tools out of the box like caching.
+
+This ADR serves to lay out pros and cons of using GraphQL and Apollo in the Settings Redesign project as an alternative to hitting our conventional REST endpoints.
+
+¹Apollo also offers Apollo Graph Manager and Apollo Federation which are paid services, [read more from their docs](https://www.apollographql.com/docs/intro/platform/). We do not need to use these to use GQL with Apollo server or Apollo client.
+
+## Decision Drivers
+
+- Performance
+- Easy API setup and React integration
+- Clarity around "what's happening?" and data received from the server
+- Development speed, both around initial setup and as new features roll out
+
+## Considered Options
+
+- Option A - Send standard CRUD requests to our REST endpoints
+- Option B - Layer GraphQL and Apollo on top of our REST architecture, making direct database calls for basic operations
+- Option C - Full GraphQL and Apollo integration with direct database calls
+
+## Decision Outcome
+
+Chosen option: "B - Layer GraphQL on top of our REST architecture", because:
+
+- GQL seems to be the ideal solution to the heavy number of network requests on the main Settings page where around 17 different requests are sent to our REST API, and a lot of data received is superfluous. GQL moves us towards a performant solution that consolidates these initial requests and reduces overfetching in other requests.
+- Allowing developers to query for and expect only exactly what is needed with end-to-end typing and a more "declarative" way of thinking towards data requirements makes it painfully clear what's being sent and received between client and server.
+- The React Settings application will only have to manage interacting with a single endpoint and Apollo Client makes managing the data in our React app straightforward.
```

Comment:
This is not a very strong _argument_ in favor, more of a nice side-effect.

Reply:
I think the GraphQL argument here is that it moves the data schemas & requirements more into the client, closer to where the data's used. The classic server-side API pattern is that the endpoints define what they return and the clients request and rework the data as necessary.

Reply:
Thanks for comments from both of you in this section, slightly rephrasing a couple of points.